File: java/rest-simple/src/main/java/com/example/restsimple/repository/UnitOfWorkImpl.java
package com.example.restsimple.repository;

import org.springframework.stereotype.Service;

import org.springframework.transaction.annotation.Transactional;

@Service
public class UnitOfWorkImpl implements UnitOfWork {

    @Override
    @Transactional
    public <T> T executeInTransaction(TransactionCode<T> code) {
        return code.apply();
    }
}

File: kotlin/rest-simple-exercise/src/main/kotlin/com/example/restsimple/repository/UnitOfWorkImpl.kt
package com.example.restsimple.repository

import com.example.restsimple.model.Student
import org.springframework.stereotype.Service
import javax.transaction.Transactional

@Service
class UnitOfWorkImpl : UnitOfWork {
    @Transactional
    override fun <T> executeInTransaction(code: () -> T): T? {
        return code.invoke()
    }
}
